api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -834,3 +834,50 @@ message RequestIDExtension {
   // Request ID extension specific configuration.
   google.protobuf.Any typed_config = 1;
 }
+
+// If this is used, the cluster will only operate on one of the possible upstream protocols (HTTP/1.1, HTTP/2).
+// If :ref:`http2_protocol_options <envoy_api_field_config.cluster.v3.Cluster.http2_protocol_options>` are
+// present, HTTP2 will be used, otherwise HTTP1.1 will be used.
+message ExplicitHttpConfig {
+  oneof protocol_config {
+    config.core.v3.Http1ProtocolOptions http_protocol_options = 1;
+
+    config.core.v3.Http1ProtocolOptions http2_protocol_options = 2;
+  }
+}
+
+// If this is used, the cluster can use either of the configured protocols, and
+// will use whichecer protocol was used by the downstream connection.
+message UseDownstreamHttpConfig {
+  config.core.v3.Http1ProtocolOptions http_protocol_options = 1;
+
+  config.core.v3.Http1ProtocolOptions http2_protocol_options = 2;
+}
+
+// If this is used, Envoy will negotiate ALPN to determine if HTTP/1 or HTTP/2 should be used.
+message AlpnHttpConfig {
```

Comment:
Should we add a string "expected ALPN" field which defaults to "h2,http/1.1" and then we can match that against what is actually configured on the transport socket if TLS, warn if no TLS, etc.?

Reply:
@snowp thoughts on this?
Previously we'd allowed the transport to override the "default" ALPN, and the cluster H1/H2 decoration was the fallback.

Reply:
I think overriding is fine, it's just a message about "expected". I think this would just allow for better error messages and make it super clear to the user what they are asking for? This will increase usability a massive amount over what we have today.

Reply:
I think using AlpnHttpConfig with a non-ALPN transport socket should just hard fail if I can make it happen.

The thing I don't follow about your requested expected_alpn is I'm not sure if you're asking about the expected alpn from the transport socket, the expected one the cluster is going to do, or some glomming of the two.  Do we expect that folks will explicitly configure transport socket ALPN to the same thing their cluster is doing?  I thought the point of transport was to allow the user to configure custom-something, then the cluster would do sensible failover so my canonical config for H2 ALPN would be putting nothing in the transport alpn field, then letting the cluster do H2,h1

Reply:
So, today here is what the user has to do if they want to use H2 and ALPN to upstream:
1) Configured http2_protocol_options to turn on h2
2) Know to configure "h2" ALPN for upstream TLS ALPN, assuming the upstream server expects that.

It's extremely common that users don't do either (1) or (2) correctly, leading to a lot of confusion and support queries.

In terms of "expected" I was suggesting we do something like:

1) `string expected_configured_transport_alpn = 1; // defaults to "h2,http/1.1" if not set.`
2) Check ^ against actual configured ALPN on transport and fails if they don't match.

This covers the default case that almost everyone wants when using this new pool type and also allows overriding it if ther is some reason for the user to do so (which would be super advanced).

Does that make sense? wdyt?

Reply:
Ah, I think I see why you thought this was helpul!

I'd planned on hard-coding the cluster's "failover" ALPN to h2/h1 for Alpn config, regardless of it HTTP/1 and HTTP/2 configs were present.  So by using AlpnHttpConfig, you do ALPN for h2,h1.  The way cluster-alpn works, it automatically does ALPN even if the transport ALPN is blank. If you stick in an override in transport config, it'll be used, but the fallback is still whatever the cluster config says (so h2.http/1.1).  By default, with no transport ALPN it'd still just do "the right thing"

Unless you want to ALPN http2 only, which there would be no way to configure for what I outlined.

So really the question is if there's value in an ALPN pool which only does h2 or only does h1?

If there is, we should go with requiring explicit http1/http2 config and checking they match expected alpn as you suggested.  If we think ALPN implies http2+http1, we can go with what I had planned.

Reply:
> If there is, we should go with requiring explicit http1/http2 config and checking they match expected alpn as you suggested. If we think ALPN implies http2+http1, we can go with what I had planned.

Ah OK, thanks, this makes sense. IMO we start with just h2,http/1.1 support and always add more config later if needed. I'm just trying to reduce the config the user needs in the common case and this sounds like it does it!